Lesson8

artursmolin opened the pull request:
1. Реализуйте все стандартные действия вместе с представлениями (видами) для ресурса вопросов:
Список вопросов теста должен выводиться на той же странице, что и сам тест
Используйте фрагменты представлений там, где это уместно
Для того чтобы метод form_with смог сформировать верный URL используйте вызов form_with model: [@parent, @child]
В представлениях должны быть ссылки для навигации между страницами приложения

2. Добавьте футер в котором будет выводиться текущий год, имя автора, название проекта (в виде ссылки на GitHub-репозиторий проекта), его краткое описание и ссылка с текстом "Учебный проект в онлайн-школе Thinknetica", ведущая на адрес thinknetica.com

3. Напишите хэлперы:
Для заголовка формы создания и редактирования вопроса. Это должен быть один метод question_header который в форме создания вопроса будет выводить Create New <Test Title> Question, а в форме редактирования Edit <Test Title> Question
Для вывода текущего года в футере
Для формирования ссылки на GitHub репозиторий: github_url(author, repo)

4. Добавьте в представление списка тестов количество вопросов для каждого теста

artursmolin committed:
commit first

1. Реализуйте все стандартные действия вместе с представлениями (видами) для ресурса вопросов:
Список вопросов теста должен выводиться на той же странице, что и сам тест
Используйте фрагменты представлений там, где это уместно
Для того чтобы метод form_with смог сформировать верный URL используйте вызов form_with model: [@parent, @child]
В представлениях должны быть ссылки для навигации между страницами приложения

2. Добавьте футер в котором будет выводиться текущий год, имя автора, название проекта (в виде ссылки на GitHub-репозиторий проекта), его краткое описание и ссылка с текстом "Учебный проект в онлайн-школе Thinknetica", ведущая на адрес thinknetica.com

3. Напишите хэлперы:
Для заголовка формы создания и редактирования вопроса. Это должен быть один метод question_header который в форме создания вопроса будет выводить Create New <Test Title> Question, а в форме редактирования Edit <Test Title> Question
Для вывода текущего года в футере
Для формирования ссылки на GitHub репозиторий: github_url(author, repo)

4. Добавьте в представление списка тестов количество вопросов для каждого теста

artursmolin committed:
+ nav + quest/form + footer + quest/edit,new

artursmolin committed:
edit in questions

fix

artursmolin committed:
new fix

header new, edit question fix
add show.html.erb to questions

artursmolin committed:
fix repo helper

fix year + github repo

artursmolin committed:
fix

fix

max-underthesun commented on test_guru/app/controllers/questions_controller.rb:
```diff
@@ -10,23 +10,38 @@ def index
   end
```
ТЗ: "Список вопросов теста должен выводиться **на той же странице, что и сам тест**"

другими словами - в этом контроллере этот экшен не нужен, индекс вопросов конкретного теста должен выводиться на странице этого теста (TestsController #show и соответствующая вьюха) , как у тебя и сделано
будешь убират экшен - не забудь убрать маршрут

max-underthesun commented on test_guru/app/controllers/questions_controller.rb:
```diff
@@ -10,23 +10,38 @@ def index
   def show
-    render inline: "ID: <%= @question.id %>, Question: <%= @question.body %>, Test_ID: <%= @question.test_id %>"
+
+  end
```
пустые экшены рекомендуется записывать в виде
```ruby
def show; end
```

max-underthesun commented on test_guru/app/controllers/questions_controller.rb:
```diff
@@ -1,6 +1,6 @@
-  before_action :find_test, only: [:index, :create]
-  before_action :find_question, only: [:destroy, :show]
+  before_action :find_test, only: [:index, :create, :new]
+  before_action :find_question, except: [:index, :new, :create]
```
в большинстве случаев лучше использовать "онли" - это более явный вариант, чем "эксепт", который стоит использовать если уж действительно все восемнадцать кроме одного или двух

max-underthesun commented on test_guru/app/controllers/questions_controller.rb:
```diff
@@ -10,23 +10,38 @@ def index
   def destroy
     @question.destroy
-    render inline: "Question - ID: <%= @question.id %>, Question: <%= @question.body %>, Test_ID: <%= @question.test_id %> has been deleted!"
+    redirect_to test_path(test.test_id)
```
что это за локальная переменная `test` - она нигде не определена
следующие комментарии - варианты кода, когда разберешься с переменной, т.е они используют пока несуществующую переменную, но предлагают тебе более удобный синтаксис

max-underthesun replied:
```suggestion
    redirect_to test_path(test)
```

max-underthesun replied:
```suggestion
    redirect_to test
```

max-underthesun commented on test_guru/app/views/tests/show.html.erb:
```diff
@@ -0,0 +1,30 @@
+  <tr>
+    <th>ID</th>
+    <th align="left">Questions</th>
+  </tr>
```
заголовка 2, а поля в парешеле `_question` - 4

max-underthesun commented on test_guru/app/views/shared/_footer.html.erb:
```diff
@@ -0,0 +1,8 @@
+<p>About: Проект TestGuru это замечательная образовательная платформа<br>
+  для создания и прохождения тестов по различным темам.<br>
+Разработка проекта ведется на Ruby on Rails.</p>
+<p>"Учебный проект в онлайн-школе <a href='thinknetica.com'>Thinknetica</a>"</p>
```
erb файл - лучше все-таки использовать хелперы рельс (`link_to`)

max-underthesun replied:
ссылки на внешние ресурсы правильно открывать в новом окне

max-underthesun commented on test_guru/app/helpers/application_helper.rb:
```diff
@@ -1,2 +1,10 @@
 module ApplicationHelper
+  def today
+     Date.today.year
+  end
```
в рельсах рекомендуется использовать `Time.current` вместо `Time.now` (`Time.new`) и `Date.current` вместо `Date.today`
`.current` возвращает текущий момент времени с учетом настройки часового пояса в приложении

max-underthesun reviewed (commented):
не нашел выполнения вот этой части ТЗ:

Напишите хэлперы:
Для заголовка формы создания и редактирования вопроса. Это должен быть один метод question_header который в форме создания вопроса будет выводить Create New <Test Title> Question, а в форме редактирования Edit <Test Title> Question

artursmolin committed:
fix

fix

max-underthesun commented on test_guru/app/helpers/questions_helper.rb:
```diff
@@ -1,4 +1,12 @@
 module QuestionsHelper
   def index
   end
```
а это что?

max-underthesun commented on test_guru/app/views/questions/edit.html.erb:
```diff
@@ -0,0 +1,15 @@
+<h1>Edit Question <%= @question.id %></h1>
```
здесь тоже нужно использовать хелпер, он для этого написан

artursmolin replied:
тут почему-то не запулилось изменение

max-underthesun commented on test_guru/app/views/questions/edit.html.erb:
```diff
@@ -0,0 +1,15 @@
+<h1>Edit Question <%= @question.id %></h1>
+
+<%= form_with model: @question, local: true do |form| %>
```
а зачем здесь дублировать форму? так же как и в "нью" здесь нужно рендерить паршел с формой

max-underthesun commented on test_guru/app/controllers/questions_controller.rb:
```diff
@@ -1,32 +1,42 @@
-    @questions = @test.questions
-    render json: @questions
-  end
+  def index; end
```
зачем здесь "индекс" вообще, даже пустой?
он не используется, вьюхи нет, значит и экшен нужно убрать
и в прошлый раз я просил тебя убрать маршрут, тоже не сделано

artursmolin replied:
"пропиздоглазил"

max-underthesun commented on test_guru/app/controllers/questions_controller.rb:
```diff
@@ -1,32 +1,42 @@
-  def new; end
+  def edit
+
+  end
```
https://github.com/drujok/Rails/pull/13#discussion_r281727449

max-underthesun commented on test_guru/app/controllers/questions_controller.rb:
```diff
@@ -1,32 +1,42 @@
     @question.destroy
-    render inline: "Question - ID: <%= @question.id %>, Question: <%= @question.body %>, Test_ID: <%= @question.test_id %> has been deleted!"
+    redirect_to tests_path
+  end
```
предыдущий вариант с редиректом на страницу теста был более удачный - это же удаление вопроса из конкретного теста, зачем потом редирект на индекс тестов
нужно только разобраться откуда взять объект теста

artursmolin committed:
fix

добавил форм на эдите в вопросах
убрал путь индекса
поправил хелпер
сделал редирект на тест

max-underthesun commented on test_guru/app/controllers/questions_controller.rb:
```diff
@@ -1,32 +1,39 @@
+    test = Test.find(@question.test_id)
     @question.destroy
-    render inline: "Question - ID: <%= @question.id %>, Question: <%= @question.body %>, Test_ID: <%= @question.test_id %> has been deleted!"
+    redirect_to test
```
сложновато получилось
переменная "вопрос" в памяти остается после "дестроя", ей можно пользоваться
также у нас есть ассоциации в моделях
поэтому "упрощенный вариант":
```ruby
@question.destroy
redirect_to @question.test
```

artursmolin committed:
fix q_controller

fix def destroy
